zipline/utils/calendars/cme_exchange_calendar.py

```diff
@@ -0,0 +1,374 @@
+    special_closes_calendars = []
+
+    holidays_adhoc = chain(
+        September11Closings,
+        HurricaneSandyClosings,
+        USNationalDaysofMourning,
+    )
+
+    special_opens_adhoc = ()
+    special_closes_adhoc = []
+
+    @property
+    def name(self):
+        """
+        The name of this exchange calendar.
+        E.g.: 'NYSE', 'LSE', 'CME Energy'
+        """
+        return 'CME'
+
+    @property
+    def tz(self):
+        """
+        The native timezone of the exchange.
+
+        SD: Not clear that this needs to be exposed.
+        """
+        return self.native_timezone
+
+    def is_open(self, dt):
+        """
+        Parameters
+        ----------
+        dt : Timestamp
+
+        Returns
+        -------
+        bool
+            True if  exchange is open at the given dt, otherwise False.
+        """
+        # Retrieve the exchange session relevant for this datetime
+        session = self.session_date(dt)
+        # Retrieve the opens and closes for this exchange session
+        o_and_c_df = self.open_and_close(session)
+        # Is @dt within the trading hours for this exchange session
+        for index, row in o_and_c_df.iterrows():
+            if row['market_open'] <= dt and dt <= row['market_close']:
+                return True
+        return False
+
+    def trading_days(self, start, end):
+        """
+        Calculates all of the exchange sessions between the given
+        start and end.
+
+        SD: Presumably @start and @end are UTC-canonicalized, as our exchange
+        sessions are. If not, then it's not clear how this method should behave
+        if @start and @end are both in the middle of the day.
+
+        Parameters
+        ----------
+        start : Timestamp
+        end : Timestamp
+
+        Returns
+        -------
+        DatetimeIndex
+            A DatetimeIndex populated with all of the trading days between
+            the given start and end.
+        """
+        return self.schedule.index[start:end]
+
+    def open_and_close(self, session):
+        """
+        Given a UTC-canonicalized date, returns a tuple of timestamps of the
+        open and close of the exchange session on that date.
+
+        SD: Can @date be an arbitrary datetime, or should we first map it to
+        and exchange session using session_date. Need to check what the
+        consumers expect. Here, I assume we need to map it to a session.
+
+        Parameters
+        ----------
+        session : Timestamp
+            The UTC-canonicalized session whose open and close are needed.
+
+        Returns
+        -------
+        (Timestamp, Timestamp)
+            The open and close for the given date.
+        """
+        # Generalised logic for the case of trading pauses.
+        # Note: this logic is ~3-4 times slower than that used for the NYSE
+        # (pass a list, to ensure we get a DataFrame returned)
+        return self.schedule.loc[[session]]
```

Comment:
This is supposed to be returning a tuple of Timestamps, instead of a dataframe. There's a `_get_open_and_close` helper method of `NYSEExchangeCalendar` that's general enough that it seems like it would be useful to share through `ExchangeCalendar`.